lambda/Scheduler/scheduler.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+
+const co = require('co');
+const _ = require('lodash');
+
+const awsFactory = require('./services/aws');
+const emFactory = require('./services/em');
+
+const scheduling = require('./scheduling');
+const reporting = require('./presentation/reporting');
+
+function createScheduler(config) {
+
+  let em = emFactory.create(config.em);
+  let ec2 = awsFactory.create(config.aws).ec2;
+
+  function doScheduling () {
+    return co(function*() {
+
+      let instances = yield getInstancesWithEnvironments(config.limitToEnvironment);
+
+      let scheduledActions = scheduledActionsForInstances(instances);
+      let actionGroups = groupActionsByType(scheduledActions);
+
+      let changeResults = yield performChanges(actionGroups);
+
+      return reporting.createReport({ actionGroups, changeResults }, config.listSkippedInstances);
+
+    }).catch(err => {
+
+      return {
+        success: false,
+        error: printableError(err),
+      };
+
+    });
+  }
+
+  function printableError(err) {
+    if (err instanceof Error) {
+      return {
+        message: err.message,
+        stack: err.stack
+      };
+    }
+
+    return err;
+  }
+
+  function getInstancesWithEnvironments(environmentFilter) {
+
+    return Promise.all([ec2.getAllInstances(), em.getAllEnvironments()]).then(data => {
+
+      let allInstances = data[0];
+      let environments = buildEnvironmentIndex(data[1]);
+
+      let instances = [];
+
+      allInstances.forEach(instance => {
+        let environmentName = getInstanceTagValue(instance, 'environment');
+
+        if (environmentMatchesFilter(environmentName, environmentFilter)) {
+          instance.Environment = findEnvironmentByName(environments, environmentName);
+          instances.push(instance);
+        }
+      });
+
+      return instances;
+
+    });
+
+  }
+
+  function environmentMatchesFilter(environmentName, environmentFilter) {
+
+    if (!environmentFilter) return true;
```

Comment:
could we have more definite conditions? like `environmentFilter !== undefined` or `environmentFilter !== null` ?
https://github.com/airbnb/javascript#comparison--shortcuts